core/src/test/java/org/elasticsearch/threadpool/ScheduleWithFixedDelayTests.java

```diff
@@ -0,0 +1,287 @@
+        // nothing should be done with the threadpool on creation
+        verifyZeroInteractions(threadPool);
+
+        Cancellable cancellable = reschedulingRunnable.scheduleFirstRun();
+        verify(threadPool, times(1)).schedule(delay, Names.GENERIC, reschedulingRunnable);
+
+        // call run to mimic scheduler starting the runnable
+        RuntimeException runtimeException = expectThrows(RuntimeException.class, reschedulingRunnable::run);
+        assertEquals(null, runtimeException.getMessage());
+
+        // verify that there were no more interactions with the threadpool such as scheduling again
+        verifyNoMoreInteractions(threadPool);
+
+        assertTrue(cancellable.isCancelled());
+    }
+
+    public void testDoesNotRescheduleUntilExecutionFinished() throws Exception {
+        final TimeValue delay = TimeValue.timeValueMillis(100L);
+        final CountDownLatch startLatch = new CountDownLatch(1);
+        final CountDownLatch pauseLatch = new CountDownLatch(1);
+        ThreadPool threadPool = mock(ThreadPool.class);
+        final Runnable runnable = () ->  {
+            startLatch.countDown();
+            try {
+                pauseLatch.await();
+            } catch (InterruptedException e) {
+                Thread.currentThread().interrupt();
+            }
+        };
+        ReschedulingRunnable reschedulingRunnable = new ReschedulingRunnable(runnable, delay, Names.GENERIC, threadPool);
+        // nothing should be done with the threadpool on creation
+        verifyZeroInteractions(threadPool);
+
+        reschedulingRunnable.scheduleFirstRun();
+        Thread runThread = new Thread() {
+            @Override
+            public void run() {
+                reschedulingRunnable.run();
+            }
+        };
+        runThread.start();
+        startLatch.await();
+
+        // this call was made during scheduleFirstRun
+        verify(threadPool, times(1)).schedule(delay, Names.GENERIC, reschedulingRunnable);
+
+        // un-pause
+        pauseLatch.countDown();
+        runThread.join();
+
+        // validate schedule called again
+        verify(threadPool, times(2)).schedule(delay, Names.GENERIC, reschedulingRunnable);
+    }
+
+    public void testScheduleAfterCancel() throws Exception {
+        final TimeValue delay = TimeValue.timeValueMillis(100L);
+        final AtomicInteger counter = new AtomicInteger(0);
+        ThreadPool threadPool = mock(ThreadPool.class);
+        Runnable countingRunnable = counter::incrementAndGet;
+        ReschedulingRunnable reschedulingRunnable = new ReschedulingRunnable(countingRunnable, delay, Names.GENERIC, threadPool);
+
+        Cancellable cancellable = reschedulingRunnable.scheduleFirstRun();
+        if (randomBoolean()) {
+            reschedulingRunnable.run();
+        }
+        cancellable.cancel();
+
+        IllegalStateException exception = expectThrows(IllegalStateException.class, reschedulingRunnable::scheduleFirstRun);
```

Comment:
schedule first run should be called twice, regardless of cancellation right? can we check for that using an assertion? we don't really have to test it as there is no way for the user to achieve this when the class is not exposed.